hwtracer/src/decode/ykpt/mod.rs

```diff
@@ -97,126 +280,355 @@ impl<'t> YkPTBlockIterator<'t> {
                             }
-                        };
-                        self.cur_loc = ObjLoc::MainObj(target_off);
-                        return Ok(Some(self.lookup_block_from_main_bin_offset(target_off)));
+                        }
+                        SuccessorKind::Conditional {
+                            taken_target,
+                            not_taken_target,
+                        } => {
+                            // If we don't have any TNT choices buffered, get more.
+                            if self.tnts.is_empty() {
+                                self.seek_tnt()?;
+                            }
+
+                            // Find where to go next based on whether the trace took the branch or not.
+                            //
+                            // The `unwrap()` is guaranteed to succeed because the above call to
+                            // `seek_tnt()` has populated `self.tnts()`.
+                            let target_off = if self.tnts.pop_front().unwrap() {
+                                *taken_target
+                            } else {
+                                if let Some(ntt) = not_taken_target {
+                                    *ntt
+                                } else {
+                                    // Divergent control flow.
+                                    todo!();
+                                }
+                            };
+                            self.cur_loc = ObjLoc::MainObj(target_off);
+                            return self.lookup_block_from_main_bin_offset(target_off);
+                        }
+                        SuccessorKind::Return => {
+                            if self.is_return_compressed()? {
+                                // This unwrap cannot fail if the CPU has implemented compressed
+                                // returns correctly.
+                                let off = match self.comprets.pop().unwrap() {
+                                    CompRetAddr::AfterCall(off) => off + 1,
+                                    CompRetAddr::VAddr(vaddr) => {
+                                        let (obj, off) = self.vaddr_to_off(vaddr)?;
+                                        debug_assert!(obj == *SELF_BIN_PATH);
+                                        off
+                                    }
+                                };
+
+                                self.cur_loc = ObjLoc::MainObj(off + 1);
+                                return self.lookup_block_from_main_bin_offset(off + 1);
+                            } else {
+                                // A regular uncompressed return that relies on a TIP update.
+                                //
+                                // Note that `is_return_compressed()` has already updated
+                                // `self.cur_loc()`.
+                                let ret = match self.cur_loc {
```

Comment:
Maybe:

```rust
match ... {
  ObjLoc::MainObj(off) => return ...,
  _ => return Block::Unknown
}
```

?

Reply:
Fixed in a466b1e0bdc4a1769572b886c98264e8f206661d